internal/cuegen/component.go

```diff
@@ -0,0 +1,92 @@
+package cuegen
+
+import (
+	"fmt"
+
+	"cuelang.org/go/cue/ast"
+	"cuelang.org/go/cue/token"
+	"github.com/benthosdev/benthos/v4/internal/docs"
+)
+
+type componentOptions struct {
+	collectionIdent  *ast.Ident
+	disjunctionIdent *ast.Ident
+
+	canLabel      bool
+	canPreProcess bool
+}
+
+func doComponents(specs []docs.ComponentSpec, opts *componentOptions) ([]ast.Decl, error) {
+	fields := make([]interface{}, 0, len(specs))
+
+	for _, v := range specs {
+		field, err := doComponentSpec(v)
+		if err != nil {
+			return nil, fmt.Errorf("failed to generate cue type for component: %s: %w", v.Name, err)
+		}
+		fields = append(fields, field)
+	}
+
+	var addons []interface{}
+	if opts.canLabel {
+		addons = append(
+			addons,
+			ast.NewIdent("label"),
+			token.OPTION,
+			ast.NewIdent("string"),
+		)
+	}
+
+	if opts.canPreProcess {
+		addons = append(
+			addons,
+			ast.NewIdent("processors"),
+			token.OPTION,
+			ast.NewList(&ast.Ellipsis{Type: ast.NewIdent("#Processor")}),
+		)
+	}
+
+	decls := []ast.Decl{
+		&ast.Field{
+			Label: opts.collectionIdent,
+			Value: ast.NewStruct(fields...),
+		},
+		&ast.Field{
+			Label: opts.disjunctionIdent,
+			Value: ast.NewCall(ast.NewIdent("or"), ast.NewList(&ast.Comprehension{
```

Comment:
So I wanted to build various disjunctions dynamically, based the available components for a given category (inputs, processors, outputs, ...). The cleanest way I've found to do this is to create a struct of all components of that category and then to you use a field comprehension and the Cue built-in function `or` to create the disjunction. You can try this following snippet in the Cue playground to see it in action:

```cue
#AllInputs: {
	broker: {
		copies?: int
		inputs?: [...null | #Input]
		batching?: {
			count?:     int
			byte_size?: int
			period?:    string
			check?:     string
			processors?: [...null | _]
		}
	}
	csv: {
		paths?: [...string]
		parse_header_row?: bool
		delimiter?:        string
		batch_count?:      int
	}
	discord: {
		channel_id:   string
		bot_token:    string
		poll_period?: string
		limit?:       int
		cache:        string
		cache_key?:   string
		rate_limit?:  string
	}
	dynamic: {
		inputs?: {
			[string]: null | #Input
		}
		prefix?: string
	}
	file: {
		paths?: [...string]
		codec?:            string
		max_buffer?:       int
		delete_on_finish?: bool
	}
}

#Input: or([for name, config in #AllInputs {(name): config}])

x: #Input

y: #Input & {
  	discord: {
		channel_id: 123  // FIXME: demoing that validation works. Changes to a string to verify.
		bot_token: "secret"
		cache: "whatever"
	}
}
```